sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -335,8 +339,26 @@ public void finalizeCheckpoint() {
     // set an empty list to messages when deserialize
     private void readObject(java.io.ObjectInputStream stream)
         throws IOException, ClassNotFoundException {
+      stream.defaultReadObject();
       messages = new ArrayList<>();
     }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other instanceof MqttCheckpointMark){
+        MqttCheckpointMark that = (MqttCheckpointMark)other;
+        return this.clientId.equals(that.clientId) && this.oldestMessageTimestamp.equals(that.oldestMessageTimestamp);
```

Comment:
Use Objects.equal since it is a null safe check and then you never need to worry if `clientId` or `oldestMessageTimestamp` are null.